lib: os: add sys_sem data type

wentongwu opened the pull request:
For systems with userspace, the sys_sem exist in user
    memory working as counter semaphore for user mode thread.
    The implemention of sys_sem is based on k_futex. And the
    majority of the synchronization operations are performed in
    user mode to reduce the calling of system call.

    Also add test cases for sys_sem.

Fixes: #15139.

aescolar commented:
@wentongwu If the futex PR is blocking this one, please link to it from here.

wentongwu commented:
@andrewboie could you please take a look the patch? Thanks

andrewboie commented on include/misc/sem.h:
```diff
@@ -0,0 +1,234 @@
+/**
+ * @brief Get sys_sem's futex.
+ *
+ * This routine returns the current futex's val of @a sem.
```
This documentation should NEVER mention futexes.
Explain what the API does, how it behaves.
End users are not concerned with how it works internally.
You can literally just cut and paste the documentation for k_sem.

andrewboie commented on include/misc/sem.h:
```diff
@@ -0,0 +1,234 @@
+ *                or one of the special values K_NO_WAIT and K_FOREVER.
+ *
+ * @retval 0 sys_sem taken.
+ * @retval -EBUSY Parameter address not recognized by the kernel.
```
why EBUSY? Shouldn't this be EACCESS or EINVAL?

andrewboie commented on include/misc/sem.h:
```diff
@@ -0,0 +1,234 @@
+ *
+ * @return N/A
+ */
+void sys_sem_give(struct sys_sem *sem);
```
we could add return values for failure cases

andrewboie commented on tests/kernel/semaphore/semaphore/src/main.c:
```diff
@@ -20,16 +21,16 @@
 struct timeout_info {
 	u32_t timeout;
-	struct k_sem *sema;
+	struct sys_sem *sema;
```
I don't think we should change the test to use sys_sem instead of k_sem.
On systems that support both, we need to test both.
Perhaps we could change this test to do something with function pointers so that we can test both without copying tons of code.

tyrantlizard77 reviewed (changes requested):
the documentation for the sys_sem APIs should not be describing these APIs in terms of how they work internally with futexes. depending on configuration, there might not even be a futex. you should simply copypaste the documentation for k_sem since they behave identically,

wentongwu commented:
> @wentongwu If the futex PR is blocking this one, please link to it from here.

Thanks, I will.

ioannisg commented:
Could you rebase this, now, @wentongwu since the futex support is merged?

wentongwu commented:
> Could you rebase this, now, @wentongwu since the futex support is merged?

yes, I'm doing it. Thanks

wentongwu commented:
shippable failed two kinds of reason, but not related to my change.
```
1) qemu_xtensa/tests/kernel/timer/timer_api/kernel.timer  Assertion failed at ZEPHYR_BASE/tests/kernel/timer/timer_api/src/main.c:136: test_timer_duration_period: tdata.expire_cnt == 4 is false
2) qemu_xtensa/tests/kernel/sched/schedule_api/kernel.sched.multiq  Assertion failed at ZEPHYR_BASE/tests/kernel/sched/schedule_api/src/test_sched_timeslice_reset.c:79: thread_tslice: (t <= expected_slice_max is false)
timeslice too big, expected 2010000 got 2150027
3) qemu_xtensa:tests/kernel/sched/preempt/kernel.sched.preempt  failure: timeout
4) qemu_xtensa:tests/kernel/timer/timer_api/kernel.timer  Assertion failed at ZEPHYR_BASE/tests/kernel/timer/timer_api/src/main.c:136: test_timer_duration_period: tdata.expire_cnt == 4 is false
5) In function 'sane_partition':
../../../../../../../include/misc/__assert.h:104:11: error: format '%x' expects argument of type 'unsigned int', but argument 2 has type 'uintptr_t' {aka 'const long unsigned int'} [-Werror=format=]
    printk("\t" fmt "\n", ##__VA_ARGS__);            \
6) include/misc/__assert.h:104:11: error: format '%x' expects argument of type 'unsigned int', but argument 4 has type 'uintptr_t' {aka 'const long unsigned int'} [-Werror=format=]
    printk("\t" fmt "\n", ##__VA_ARGS__);   
7) zephyr/arch/arc/core/mpu/arc_mpu_v2_internal.h:328:12: error: format '%x' expects argument of type 'unsigned int', but argument 3 has type 'uintptr_t' {aka 'long unsigned int'} [-Werror=format=]
    LOG_DBG("set region 0x%x 0x%x 0x%x",
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~
     region_index, pparts->start, pparts->size);) 
```

ioannisg commented:
@wentongwu this compilation issue is now fixed, AFAIK

ioannisg commented on include/misc/sem.h:
```diff
@@ -0,0 +1,103 @@
+
+#ifdef CONFIG_USERSPACE
+/*
+ * sys_sem behaves almost exactly like k_sem, with the added
```
Is "almost exactly" a proper expression?
@wentongwu are there any other differences with k_sem, other than the fact that sys_sem can reside in user memory?

ioannisg commented on kernel/userspace.c:
```diff
@@ -21,6 +21,7 @@
 #include <app_memory/app_memdomain.h>
 #include <misc/libc-hooks.h>
 #include <misc/mutex.h>
+#include <misc/sem.h>
```
Is this header needed in this source file?

ioannisg commented on include/misc/sem.h:
```diff
@@ -0,0 +1,103 @@
+ * for sys_sem. And the consumer of sys_sem will decrease one at every
+ * semaphore take if sys_sem's value greater than zero. But when it's
+ * value is zero, the consumer will sleep until producer's wakeup or
+ * timeout happens depended on the timeout parameter. When timeout
```
depending

ioannisg reviewed (commented):
Looks good. Going to do some testing and report here

wentongwu commented:
> Looks good. Going to do some testing and report here

@ioannisg thanks for the great help, I will change the code as your comments. If the testing found any issue, please let me know. Thanks.

ioannisg reviewed (dismissed):
Tests are passing on ARMv7-M and ARMv8-M boards, I can confirm.
Looks good, pls, address the editorial comments i had in the review.
Also, we could rename sem.h and sem.c to sys_sem.h and sys_sem.c, respectively, to match the provided API. But that's minor.

wentongwu commented:
the failed two cases not related to my patch, I don't know what's going on about CI these two days.
1) up_squared/samples/synchronization/sample.synchronization  failure: build_error
```
FAILED: ccache /opt/sdk/zephyr-sdk-0.10.1/i586-zephyr-elf/bin/i586-zephyr-elf-gcc -DBUILD_VERSION=zephyr-v1.14.0-1825-g25a29ffa1719 -DKERNEL -D_FORTIFY_SOURCE=2 -D__ZEPHYR__=1 -I../../../../../kernel/include -I../../../../../arch/x86/include -I../../../../../include -I../../../../../include/drivers -Izephyr/include/generated -I../../../../../soc/x86/apollo_lake -I../../../../../lib/libc/minimal/include -isystem /opt/sdk/zephyr-sdk-0.10.1/i586-zephyr-elf/bin/../lib/gcc/i586-zephyr-elf/8.3.0/include -isystem /opt/sdk/zephyr-sdk-0.10.1/i586-zephyr-elf/bin/../lib/gcc/i586-zephyr-elf/8.3.0/include-fixed -Os -nostdinc --imacros=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/out-3nd-pass/up_squared/samples/synchronization/sample.synchronization/zephyr/include/generated/autoconf.h --imacros=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/include/toolchain/zephyr_stdint.h -ffreestanding -fno-common -g -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wno-unused-but-set-variable -Werror=implicit-int -Werror -Wno-deprecated-declarations -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/samples/synchronization=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr=ZEPHYR_BASE -ffunction-sections -fdata-sections -mpreferred-stack-boundary=2 -mno-sse -march=silvermont -march=atom -std=c99 -MD -MT zephyr/CMakeFiles/zephyr.dir/drivers/timer/apic_timer.c.obj -MF zephyr/CMakeFiles/zephyr.dir/drivers/timer/apic_timer.c.obj.d -o zephyr/CMakeFiles/zephyr.dir/drivers/timer/apic_timer.c.obj   -c /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/timer/apic_timer.c
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/timer/apic_timer.c:6:10: fatal error: drivers/system_timer.h: No such file or directory
 #include <drivers/system_timer.h>
```

2) up_squared/samples/boards/up_squared/gpio_counter/sample.board.up_squared.gpio_counter  failure: build_error
```
FAILED: ccache /opt/sdk/zephyr-sdk-0.10.1/i586-zephyr-elf/bin/i586-zephyr-elf-gcc -DBUILD_VERSION=zephyr-v1.14.0-1825-g25a29ffa1719 -DKERNEL -D_FORTIFY_SOURCE=2 -D__ZEPHYR__=1 -I../../../../../../../kernel/include -I../../../../../../../arch/x86/include -I../../../../../../../include -I../../../../../../../include/drivers -Izephyr/include/generated -I../../../../../../../soc/x86/apollo_lake -I../../../../../../../lib/libc/minimal/include -I../../../../../../../boards/x86/up_squared -isystem /opt/sdk/zephyr-sdk-0.10.1/i586-zephyr-elf/bin/../lib/gcc/i586-zephyr-elf/8.3.0/include -isystem /opt/sdk/zephyr-sdk-0.10.1/i586-zephyr-elf/bin/../lib/gcc/i586-zephyr-elf/8.3.0/include-fixed -Os -nostdinc --imacros=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/out-3nd-pass/up_squared/samples/boards/up_squared/gpio_counter/sample.board.up_squared.gpio_counter/zephyr/include/generated/autoconf.h --imacros=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/include/toolchain/zephyr_stdint.h -ffreestanding -fno-common -g -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wno-unused-but-set-variable -Werror=implicit-int -Werror -Wno-deprecated-declarations -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/samples/boards/up_squared/gpio_counter=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr=ZEPHYR_BASE -ffunction-sections -fdata-sections -mpreferred-stack-boundary=2 -mno-sse -march=silvermont -march=atom -std=c99 -MD -MT zephyr/CMakeFiles/zephyr.dir/drivers/timer/apic_timer.c.obj -MF zephyr/CMakeFiles/zephyr.dir/drivers/timer/apic_timer.c.obj.d -o zephyr/CMakeFiles/zephyr.dir/drivers/timer/apic_timer.c.obj   -c /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/timer/apic_timer.c
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/timer/apic_timer.c:6:10: fatal error: drivers/system_timer.h: No such file or directory
 #include <drivers/system_timer.h>
```

wentongwu commented:
> Tests are passing on ARMv7-M and ARMv8-M boards, I can confirm.
> Looks good, pls, address the editorial comments i had in the review.
> Also, we could rename sem.h and sem.c to sys_sem.h and sys_sem.c, respectively, to match the provided API. But that's minor.

@ioannisg  Thanks

andrewboie commented on include/misc/sem.h:
```diff
@@ -0,0 +1,104 @@
+#ifndef ZEPHYR_INCLUDE_MISC_SEM_H_
+#define ZEPHYR_INCLUDE_MISC_SEM_H_
+
+#ifdef CONFIG_USERSPACE
```
You need an alternate implementation which maps directly onto k_sem APIs if userspace isn't enabled.

See what was done in include/misc/mutex.h

andrewboie commented on include/misc/sem.h:
```diff
@@ -0,0 +1,104 @@
+/**
+ * @brief Give a sys_sem.
+ *
+ * This routine gives @a sys_sem. The producer of sys_sem will increase
```
Please just use the same documentation as k_sem APIs.
The fact that we use atomic operations / futexes is an implementation detail and shouldn't be discussed here.

andrewboie commented on include/misc/sem.h:
```diff
@@ -0,0 +1,104 @@
+/**
+ * @brief Take a sys_sem.
+ *
+ * This routine takes @a sys_sem. This function handles the synchronization
```
See above

andrewboie commented on include/misc/sem.h:
```diff
@@ -0,0 +1,104 @@
+ */
+static inline unsigned int sys_sem_count_get(struct sys_sem *sem)
+{
+	return sem->futex.val;
```
atomic_get()?

andrewboie commented on tests/kernel/mem_protect/sys_sem/testcase.yaml:
```diff
@@ -0,0 +1,5 @@
+tests:
+  kernel.memory_protection.sys_sem:
+    filter: CONFIG_ARCH_HAS_USERSPACE
```
sys_sem needs to work with userspace disabled. remove this line.

wentongwu commented on lib/os/sem.c:
```diff
@@ -0,0 +1,51 @@
+	int ret = 0;
+
+	atomic_add(&sem->futex.val, 1);
+	ret = k_futex_wake(&sem->futex, false);
```
Currently the code considered SMP case, the multiple operations(more than one instruction) on futex should only happen on one side, take or give, because of contend even it's atomic type. So on give side,  I do k_futex_wake unconditionally. If we don't consider SMP, I can add the compare.

andrewboie replied:
Why can't we do:

```
if (atomic_add(&sem->futex.val, 1) == 0) {
   /* Futex was contended. Wake one of the sleeping threads up. */
   ret = k_futex_wake(&sem->futex, false);
}
```

andrewboie replied:
This may be worth reviewing:

https://github.com/aosp-mirror/platform_bionic/blob/master/libc/bionic/semaphore.cpp

wentongwu replied:
> Why can't we do:
> 
> ```
> if (atomic_add(&sem->futex.val, 1) == 0) {
>    /* Futex was contended. Wake one of the sleeping threads up. */
>    ret = k_futex_wake(&sem->futex, false);
> }
> ```
yes, you are right, I consider more for sys_sem_give.

andrewboie commented on lib/os/sem.c:
```diff
@@ -0,0 +1,51 @@
+	atomic_add(&sem->futex.val, 1);
+	ret = k_futex_wake(&sem->futex, false);
+
+	if (ret) {
```
What is the purpose of this check?
As far as I can tell, it hides any errors in giving the semaphore.

andrewboie commented on lib/os/sem.c:
```diff
@@ -0,0 +1,51 @@
+
+void sys_sem_init(struct sys_sem *sem)
+{
+	k_futex_init(&sem->futex);
```
atomic_clear(&sem->futex.val);

wentongwu commented:
@andrewboie could you please take a look the new patch? Thanks

andrewboie commented on lib/os/sem.c:
```diff
@@ -0,0 +1,109 @@
+void sys_sem_init(struct sys_sem *sem, unsigned int initial_count,
+		unsigned int limit)
+{
+	__ASSERT(limit != 0U, "limit cannot be zero");
```
Don't make these assertions. Return errors instead.

andrewboie commented on lib/os/sem.c:
```diff
@@ -0,0 +1,109 @@
+	do {
+		old_value = atomic_get(&sem->futex.val);
+		if (old_value >= sem->limit) {
+			break;
```
No test coverage for this case.

andrewboie commented on lib/os/sem.c:
```diff
@@ -0,0 +1,109 @@
+			return 0;
+		}
+	} else if (old_value >= sem->limit) {
+		return -EAGAIN;
```
No test coverage for this case.

andrewboie commented on lib/os/sem.c:
```diff
@@ -0,0 +1,109 @@
+	return ret;
+}
+#else
+void sys_sem_init(struct sys_sem *sem, unsigned int initial_count,
```
There's no code coverage for these non-userspace versions.
Please add another scenario to test them. You can see what I did in tests/kernel/mutex/sys_mutex

andrewboie commented on lib/os/sem.c:
```diff
@@ -0,0 +1,109 @@
+	if (old_value < 0) {
+		ret = k_futex_wake(&sem->futex, true);
+
+		if (ret > 0) {
```
Need test coverage for when ret <= 0

tyrantlizard77 reviewed (changes requested):
@wentongwu can you do a code coverage run and ensure that we have full function/line/branch coverage for this code?

I've highlighted a few.

You can generate reports with 

```
sanitycheck -n -p qemu_x86 -p mps2_an385 --coverage --coverage-platform qemu_x86 --coverage-platform mps2_an385 --gcov-tool /opt/zephyr-sdk/i586-zephyr-elf/bin/i586-zephyr-elf-gcov -j12 --enable-slow -T tests/kernel/mem_protect/sys_sem/
```

wentongwu commented:
@andrewboie have add one more test case to cover the code. Also run the coverage command, but doesn’t genareate coverage data. I will take more time to get coverage work. But I review the system semaphore code carefully, I think it should coverage most the case.

tyrantlizard77 reviewed (changes requested):
This still doesn't have full code coverage.

If you rebase to the latest tree you can generate a report with:

```
sanitycheck --coverage -p qemu_x86 -T tests/kernel/mem_protect/sys_sem/
```

Make sure you have 'lcov' installed.

wentongwu commented:
> This still doesn't have full code coverage.
> 
> If you rebase to the latest tree you can generate a report with:
> 
> ```
> sanitycheck --coverage -p qemu_x86 -T tests/kernel/mem_protect/sys_sem/
> ```
> 
> Make sure you have 'lcov' installed.

@andrewboie below is the currect result:

tests/kernel/mem_protect/sys_sem/src |   |   | 97.5 % | 198 / 203 | 100.0 % | 24 / 24 | 86.1 % | 31 / 36
-- | -- | -- | -- | -- | -- | -- | -- | --
 

wentongwu commented:
@andrewboie below is the new one, I review the code again and again, I really don't know how to improve more, is there any tool showing which line not covered  and which branch not covered.

tests/kernel/mem_protect/sys_sem/src |   |   | 98.0 % | 198 / 202 | 100.0 % | 24 / 24 | 96.9 % | 31 / 32
-- | -- | -- | -- | -- | -- | -- | -- | --
 

andrewboie commented:
>  I really don't know how to improve more, is there any tool showing which line not covered and which branch not covered.

Yes. In the coverage report, click on the filename and it will show a line-by-line breakdown.

wentongwu commented:
@andrewboie for below case I don't know how to coverage, maybe this is tool's bug

                 :            : 
      56                 :          1 : void sem_give_task(void *p1, void *p2, void *p3)
      57                 :            : {
      58                 :          1 :         sys_sem_give(&simple_sem);
      59                 :          0 : }
      60                 :            : 
      61                 :          1 : void sem_take_timeout_forever_helper(void *p1, void *p2, void *p3)
      62                 :            : {
      63                 :          1 :         k_sleep(K_MSEC(100));
      64                 :          1 :         sys_sem_give(&simple_sem);
      65                 :          0 : }
      66                 :            : 
      67                 :          1 : void sem_take_timeout_isr_helper(void *p1, void *p2, void *p3)
      68                 :            : {
      69                 :          1 :         sem_give_from_isr(&simple_sem);
      70                 :          0 : }
      71                 :            :

andrewboie commented:
> @andrewboie for below case I don't know how to coverage, maybe this is tool's bug

We don't need to worry about the test/validation code, just any new code you are putting into the kernel itself.

andrewboie commented:
@wentongwu I'm not getting the same numbers as you. I think you are trying to do coverage in the test case, and not the sys_sem implementation itself. Covering the implementation is what actually matters.

I pulled your sys_sem branch and ran:

```
sanitycheck --coverage -T tests/kernel/mem_protect/sys_sem/ -p qemu_x86
```

For lib/os/sem.c I get 71.4% branches, 97.2% lines

These figures are actually misleading since the non-userspace versions are not tested at all. Please add an alternate test case configuration to cover them, see what I did in tests/kernel/mutex/sys_mutex:

```
tests:
  sys.mutex:
    tags: kernel userspace
  sys.mutex.nouser:
    tags: kernel
    extra_configs:
      - CONFIG_TEST_USERSPACE=n
```

wentongwu commented:
sem.c |   |   | 100.0 % | 50 / 50 | 100.0 % | 4 / 4 | 90.6 % | 29 / 32
-- | -- | -- | -- | -- | -- | -- | -- | --
 

@andrewboie the not covered 3 branches need SMP support to construct the case, so far we can just cover 29/32.

andrewboie commented:
@ioannisg can you re-visit your review?

andyross commented on lib/os/sem.c:
```diff
@@ -0,0 +1,113 @@
+	int ret = 0;
+	int old_value;
+
+	do {
```
Style: Make this outer loop `while (ret == 0 || ret == -EAGAIN)`.  "do" loops are always more confusing, and by initializing ret above you've guaranteed the entry condition to the loop already.  No one needs to read to the bottom of your function to see what's going on.

wentongwu replied:
@andyross I think do { } while (); can save cpu cycles.

andyross commented on lib/os/sem.c:
```diff
@@ -0,0 +1,113 @@
+			old_value < 0 ? 1 : old_value + 1) == 0);
+
+	if (old_value < 0) {
+		ret = k_futex_wake(&sem->futex, true);
```
Why true here for the wake_all parameter?  You've incremented the semaphore by one, we only need to wake up one thread, by definition.

wentongwu replied:
@andyross I use -1(futex.val) to donate there is contend on the sys_sem, maybe there is only one waiter or many waiters on this sys_sem. For many waiters case, if just wake up one waiter, others have no chance to be woken up. And wake them all, it will sleep wait again in sys_sem_take. And for one waiter case, wake all == wake one.

For why I use -1 instead of just decrement, because when decrement to -INT_MAX, we also should consider the left bound.

andrewboie replied:
I think this is fine, BIONIC does the same thing.

andyross commented on lib/os/sem.c:
```diff
@@ -0,0 +1,113 @@
+		}
+	} while (atomic_cas(&sem->futex.val, old_value,
+			old_value < 0 ? 1 : old_value + 1) == 0);
+
```
Style: you're using this pattern repeatedly, and it's actually fairly hard to understand.  How about abstract out something like:

```c
static inline atomic_t bounded_dec(atomic_t *val, atomic_t minimum)
{
    atomic_t old_value, new_value;

    do {
           old_value = atomic_get(val);
	   new_value = val <= minimum ? minimum : val - 1;
    } while(atomic_cas(val, old_value, new_value) == 0);

    return old_value;
}

static inline atomic_t bounded_inc(atomic_t *val, atomic_t maximum)
{
    atomic_t old_value, new_value;

    do {
           old_value = atomic_get(val);
	   new_value = val >= maximum ? maximum : val + 1;
    } while(atomic_cas(val, old_value, new_value) == 0);

    return old_value;
}
```

IMHO, this would clarify your atomic loops a lot, while generating the same code.

wentongwu replied:
thanks, I will do the abstract as you suggested.

andrewboie replied:
@wentongwu please address this

wentongwu replied:
@andrewboie  update the code , please take a look.

andyross commented on include/misc/sem.h:
```diff
@@ -0,0 +1,104 @@
+ *         (0, INT_MAX] and initial_count shouldn't be greater than limit.
+ */
+int sys_sem_init(struct sys_sem *sem, unsigned int initial_count,
+		unsigned int limit);
```
Design note: do we want "limit" here?  Why?  Who uses it?  This is IMHO sort of an API wart.  It provides very little value, and has a runtime cost in both CPU cycles and memory bytes.  If we're going to invent a new semaphore API, my vote is we junk it.

andrewboie replied:
@andyross I'm confused here. limit is the value in which further attempts to take result in blocking. if I have a four count semaphore, limit=4 I don't see a way around this.

andyross replied:
It's not, it's the value at which further attempts to **give** result in blocking.  And that's a werid quirk of our API.  Normally semaphores are documented to handle arbitrary up counts, the standard definition of the give/up/V/release/whatever operation almost never talks about the possibility of it blocking.  Check the docs on linux sem_post(), win32 ReleaseSemaphore(), the wikipedia "Semaphore (Programming)" page, et. al.

I'm sure there are other OSes somewhere with this behavior, we can't be the only one.  But it's just a weird thing.  And we're carrying it around as a 4 byte per semaphore cost and a few cycles on every op, and I think it's needless.

Per another comment here, I'd prefer to see us wrap **this** API as the basis for k_sem, and put the weird legacy thing there.  But that requires removing it here.

andyross replied:
Actually, I just checked: our behavior is even weirder.  We don't, in fact, block if we hit the limit.  We just throw the up value away.  That's actually IMHO worse, as it makes the number of calls to give and take mismatched when a limit is hit.  The "writing" half of an app will expect to have woken up N threads to handle N events or whatever, but in fact the number of running threads is less due to the limit.

andrewboie replied:
>> limit is the value in which further attempts to take result in blocking. if I have a four count semaphore, limit=4 I don't see a way around this.

@andyross  I mis-read the code. Taking decrements the count. Giving increases it. If you give more than the limit, the count stays at the limit.

This all seems working as intended to me.

If we want to change the semantics of our semaphore APIs, this is not the PR to do it.

andyross replied:
It's literally a new semaphore, though.  We can have any API we want.  Why pick the old weird one except pure inertia?  Like I said, we can put the limit into the k_sem wrapper.

andrewboie replied:
> We can have any API we want. Why pick the old weird one except pure inertia? 

I wanted them to have exactly the same semantics, so it isn't difficult to merge them into a single API later.

andyross replied:
And I want limit to not be carried forward as cruft, **because** it makes it harder to merge them to a cruft-free API later. :)

But again, this isn't a -1.  I'm just saying that putting this API in here means one extra place we need to deprecate later.

andrewboie replied:
> But again, this isn't a -1. I'm just saying that putting this API in here means one extra place we need to deprecate later.

I at least am all right with that.
k_sem has been like this for at least two years, ever since Ben did the kernel API overhaul. I'm actually kinda shocked you only just noticed this now. I don't think this PR is the place to redesign semaphores, not this late in the dev cycle for 2.0. I also think it's more trouble than it's worth to remove 'limit' from k_sem.

andyross commented on lib/os/sem.c:
```diff
@@ -0,0 +1,113 @@
+				break;
+			}
+		} while (atomic_cas(&sem->futex.val,
+				old_value, old_value - 1) == 0);
```
This loop could be "`old_value = bounded_dec(&sem->futex.val, -1);`"

andyross commented on lib/os/sem.c:
```diff
@@ -0,0 +1,113 @@
+
+	return ret;
+}
+#else
```
FWIW: I think this is backwards.  **This** is the cleaner, more straightforward and more futureproof synchronization primitive.  What we really want is for the legacy k_sem implementation to be reworked in terms of sys_sem, not the reverse.  And yes, I know that k_futex APIs are only defined when USERSPACE is, but that's a mistake too -- they're just minimal wrappers around an atomic_t and a wait_q and there's no particular reason we couldn't make them work in kernel code too.

Not an issue for this patch, but long term I think we want just one set of shared memory synchronization primitives.

andrewboie replied:
> they're just minimal wrappers around an atomic_t and a wait_q and there's no particular reason we couldn't make them work in kernel code too.

One thing at a time, this is brand new immature code and we need to make sure it is stable and correct before we go in and throw out our tried-and-tested k_sem implementation

Also to be honest: the whole point of this stuff is to avoid system calls. We have no system calls if user mode is disabled.

andrewboie replied:
> there's no particular reason we couldn't make them work in kernel code too.

Actually there is: there is a dependency on the kernel object tracking for the futexes to instantiate the backing waitqueue object.

We could conditionally just add the waitq to the futex data struct if userspace is disabled, but then we have to worry about waitq initialization; currently futexes do not need to be initialized before they are used.

What I'm trying to say here is that this is not as simple as you think. I still want to do this, but not until at least 2.1

andyross reviewed (approved):
Can't see anything wrong, though I had to spend a lot of time working out how all those do-loops were operating.  One strong suggestion about a way to clarify what's happenening (if nothing else, a comment to the effect of "this is a bounded increment", etc.. would have been helpful).

A few design notes too.

andrewboie commented on include/misc/sem.h:
```diff
@@ -0,0 +1,104 @@
+ * @brief public sys_sem APIs.
+ */
+
+#ifndef ZEPHYR_INCLUDE_MISC_SEM_H_
```
@wentongwu include/misc/ is deprecated.
Need to move this to include/sys/ instead.

wentongwu replied:
@andrewboie done

wentongwu committed:
lib: os: add sys_sem data type

For systems with userspace, the sys_sem exist in user memory working
as counter semaphore for user mode thread. The implemention of sys_sem
is based on k_futex. And the majority of the synchronization operations
are performed in user mode to reduce the calling of system call.
And for systems without userspace enabled, sys_sem behaves like k_sem.

Fixes: #15139.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
tests: add test cases for sys_sem

add test cases for sys_sem data structure.

Fixes: #15139.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>
